blogposts/2022/css-modules-migration.md

```diff
@@ -0,0 +1,210 @@
+3. Make changes to the AST where appropriate.
+4. Regenerate the source file based on the new AST.
+
+<figure>
+  <img src="https://storage.googleapis.com/sourcegraph-assets/blog/code-insights-ga-blogs/abstract-syntax-tree.svg" alt="Abstract syntax tree" class="no-shadow">
+  <figcaption>High-level visual representation of the codemod.</figcaption>
+</figure>
+
+On the technical side: most of the changes required for CSS Modules migration are mechanical and can be automated with the information available in CSS and Typescript abstract syntax trees. But most importantly, the codemod approach allows us to overcome all the challenges we encountered working on the migration manually:
+
+1. ⚡️Instead of pouring time into mindless repetition of migrating global styles to CSS modules, we put effort into building a robust script that does that for us. It is a much more engaging task because it gives the team knowledge and experience that can be used for future migrations.
+2. 🥷 Having a way to migrate 95% of use cases automatically greatly reduces the time commitment needed and allows the platform team to complete the refactoring behind the scenes without asking product teams for help.
+3. 🏎 Faster migration means that the codebase stays inconsistent for much less time.
+
+## Codemod implementation
+
+To programmatically migrate global styles to CSS modules, codemod script does three major things:
+
+### 1. Find pairs of files to transform.
+
+Finding a React component file with the corresponding global-styles file is easy in the Sourcegraph codebase because React components have colocated styles with a predictable filename. E.g., styles for `Button.tsx` are defined in `Button.scss` in the same folder.
+
+### 2. Convert global styles file into a CSS module.
+
+It’s a multistep operation:
+
+- First, get rid of the BEM notation used for CSS selectors and remove redundant nesting required to ensure selector uniqueness.
+
+```css
+.insights-dashboard {
+  flex: none;
+
+  &__wrapper {
+    display: flex;
+  }
+}
+```
+
+⬇
+
+```css
+.insights-dashboard {
+  flex: none;
+}
+
+.wrapper {
+  display: flex;
+}
+```
+
+- To achieve this, we used [PostCSS](https://github.com/postcss/postcss)—a ​tool for transforming styles with JS plugins. The resulting script is similar to [the postcss-nested plugin](https://github.com/postcss/postcss-nested), with some adjustments to remove the BEM notation.
+- Save a list of CSS Module classes in memory for later reference in the React component. It is done by using utility functions provided by [the css-modules-loader-core package](https://github.com/css-modules/css-modules-loader-core). They give us a list of classes available in a CSS Module that we can map ourselves to global CSS classes that we just transformed. For our toy example, it would look like this:
+
+```css
+const classMapping = {
+  'insights-dashboard':'insights-dashboard','insights-dashboard__wrapper' : 'wrapper';
+}
+```
```

Comment:
Same here, is this right?
<img width="869" alt="image7" src="https://user-images.githubusercontent.com/82806852/155334143-26dc97d5-aee9-46a3-ac5f-1c764d5b4be2.png">